### ISSUE TITLE
"Non-transitive" promotion rules with `One` and `Zero`.

### ISSUE THREAD

**Issue by KristofferC** [Jutho/VectorInterface.jl#17]
The following calls have different results depending on the order of the arguments to `promote_type`:

```julia
import VectorInterface as VI
promote_type(VI.One, VI.Zero, Float64) # -> StackOverflowError
promote_type(VI.One, Float64, VI.Zero) # -> Float64
promote_type(Float64, VI.One, VI.Zero) # -> Float64
```

Due to some changes in internals, in 1.10 this instead results in:

```julia
promote_type(VI.One, VI.Zero, Float64) # -> Float64
promote_type(VI.One, Float64, VI.Zero) # -> Float64
promote_type(Float64, VI.One, VI.Zero) # -> StackOverflowError
```

Due to chance, the matmul code happen to call the first combination of this:

https://github.com/JuliaLang/julia/blob/58ed79361f4faad0da8d96aeaffbd8d75ebbb386/stdlib/LinearAlgebra/src/matmul.jl#L500

which means it works in 1.10 but not in 1.11 (here is the PkgEval error for an upstream package https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/af49d9f_vs_08e1fc0/FiniteMPS.primary.log). In theory, the `promote` call in LinearAlgebra could be made to call the one with the number in the middle but it is a quite brittle situation.

**Comment by Jutho** [comment]
Thanks for reporting this and my apologies for the delay. I think adding
```julia
Base.promote_rule(::Type{Zero}, ::Type{One}) = Bool
```
fixes this, so I will commit this fix asap.

**Comment by Jutho** [comment]
This is fixed in the upcoming 0.4.6. Thanks again for the report.

**Comment by KristofferC** [comment]
:+1:
